HW2/P5/physics.pyx

```diff
@@ -69,17 +73,28 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+    gridx = <unsigned int> (XY[i,0] / grid_spacing)
+    gridy = <unsigned int> (XY[i,1] / grid_spacing)
+    grid_size = <unsigned int>((1.0 / grid_spacing) + 1)
+    for jx in range(gridx + 1, gridx + 3):
+        for jy in range(gridy + 1, gridy + 4- (jx - gridx)):
+            if (jx < grid_size) and (jy<grid_size) and (Grid[jx, jy]<count):
+                # avoid double-locking
+                if i != Grid[jx, jy]:
+                    acquire(&locks[Grid[jx, jy]])
+                XY2 = &(XY[Grid[jx,jy],0])
+                V2 = &(V[Grid[jx,jy],0])
+                if overlapping(XY1, XY2, R):
+                # SUBPROBLEM 4: Add locking
+                    if not moving_apart(XY1, V1, XY2, V2):
+                        collide(XY1, V1, XY2, V2)
+
+                    # give a slight impulse to help separate them
+                    for dim in range(2):
+                        V2[dim] += eps * (XY2[dim] - XY1[dim])
+                if i != Grid[jx, jy]:
+                    release(&locks[Grid[jx, jy]])
+    release(&locks[i])
```

Comment:
Another potential source of overhead - you are holding on to locks[i] for the entire loop, while other threads might be waiting for it.  You may want to try acquiring and releasing both locks only at the exact points where you need them?